crates/qpy/src/circuit_reader.rs

```diff
@@ -0,0 +1,995 @@
+    annotation_factories: &Bound<PyDict>,
+) -> PyResult<Py<PyAny>> {
+    let instruction_capacity = packed_circuit.instructions.len();
+    // create an empty circuit; we'll fill data as we go along
+    let mut circuit_data =
+        CircuitData::with_capacity(0, 0, instruction_capacity, Param::Float(0.0))?;
+
+    let annotation_handler = AnnotationHandler::new(annotation_factories);
+    let mut qpy_data = QPYReadData {
+        circuit_data: &mut circuit_data,
+        version,
+        use_symengine,
+        standalone_vars: HashMap::new(),
+        standalone_stretches: HashMap::new(),
+        vectors: HashMap::new(),
+        annotation_handler,
+    };
+
+    let annotation_deserializers_data: Vec<(String, Bytes)> = packed_circuit
+        .annotation_headers
+        .state_headers
+        .iter()
+        .map(|data| (data.namespace.clone(), data.state.clone()))
+        .collect();
+    qpy_data
+        .annotation_handler
+        .load_deserializers(annotation_deserializers_data)?;
+    let global_phase = unpack_param_from_data(
+        packed_circuit.header.global_phase_data.clone(),
+        packed_circuit.header.global_phase_type,
+        &mut qpy_data,
+        binrw::Endian::Big,
+    )?;
+    qpy_data.circuit_data.set_global_phase(global_phase)?;
+
+    add_standalone_vars(packed_circuit, &mut qpy_data)?;
+    add_registers_and_bits(packed_circuit, &mut qpy_data)?;
+
+    let custom_instructions = read_custom_instructions(py, packed_circuit, &mut qpy_data)?;
+    for instruction in &packed_circuit.instructions {
+        let inst = unpack_instruction(py, instruction, &custom_instructions, &mut qpy_data)?;
+        qpy_data.circuit_data.push(inst)?;
+    }
+
+    for (vector, initialized_params) in qpy_data.vectors.values() {
+        let vector_length = vector
+            .bind(py)
+            .call_method0("__len__")?
+            .extract::<usize>()?;
+        let missing_indices: Vec<u64> = (0..vector_length as u64)
+            .filter(|x| !initialized_params.contains(&(*x as u32)))
+            .collect();
+        if initialized_params.len() != vector_length {
+            let msg = format!(
+                "The ParameterVector: '{:}' is not fully identical to its \
+                pre-serialization state. Elements {:} \
+                in the ParameterVector will be not equal to the pre-serialized ParameterVector \
+                as they weren't used in the circuit: {:}",
+                vector.getattr(py, "name")?.extract::<String>(py)?,
+                missing_indices
+                    .iter()
+                    .map(|index| index.to_string())
+                    .collect::<Vec<_>>()
+                    .join(", "),
+                packed_circuit.header.circuit_name
+            );
+            imports::WARNINGS_WARN.get_bound(py).call1((
+                msg,
+                imports::BUILTIN_USER_WARNING.get_bound(py),
+                1,
+            ))?;
+        }
+    }
```

Comment:
Should we have something like py_methods.py for deserialization logic which heavily depends on Python, like this?